api/filters/reference_instance_filter.go

```diff
@@ -0,0 +1,174 @@
+	ctx := req.Context()
+	instanceID := req.PathParams[web.PathParamResourceID]
+
+	dbInstanceObject, err := storage.GetObjectByField(ctx, rif.repository, types.ServiceInstanceType, "id", instanceID)
+	if err != nil {
+		return next.Handle(req)
+	}
+	instance := dbInstanceObject.(*types.ServiceInstance)
+	if instance.ReferencedInstanceID == "" {
+		return next.Handle(req)
+	}
+
+	body := map[string]string{
+		instance_sharing.ReferencedInstanceIDKey: instance.ReferencedInstanceID,
+	}
+
+	return util.NewJSONResponse(http.StatusOK, body)
+}
+
+func (*referenceInstanceFilter) FilterMatchers() []web.FilterMatcher {
+	return []web.FilterMatcher{
+		{
+			Matchers: []web.Matcher{
+				web.Path(web.ServiceInstancesURL + "/**"),
+				web.Methods(http.MethodPost, http.MethodPatch, http.MethodGet),
+			},
+		},
+	}
+}
+
+func (rif *referenceInstanceFilter) handleProvision(req *web.Request, next web.Handler) (*web.Response, error) {
+	ctx := req.Context()
+	servicePlanID := gjson.GetBytes(req.Body, planIDProperty).String()
+	isReferencePlan, err := storage.IsReferencePlan(ctx, rif.repository, types.ServicePlanType.String(), "id", servicePlanID)
+	// If plan not found on provisioning, allow sm to handle the issue
+	if err == util.ErrNotFoundInStorage {
+		return next.Handle(req)
+	}
+	if err != nil {
+		return nil, err // handled by the IsReference function
+	}
+	if !isReferencePlan {
+		return next.Handle(req)
+	}
+
+	parameters := gjson.GetBytes(req.Body, "parameters").Map()
+
+	referencedInstanceID, exists := parameters[instance_sharing.ReferencedInstanceIDKey]
+	if !exists {
+		return nil, util.HandleInstanceSharingError(util.ErrMissingOrInvalidReferenceParameter, instance_sharing.ReferencedInstanceIDKey)
+	}
+
+	req.Body, err = sjson.SetBytes(req.Body, instance_sharing.ReferencedInstanceIDKey, referencedInstanceID.String())
+	if err != nil {
+		log.C(ctx).Errorf("Unable to set the instance_sharing.ReferencedInstanceIDKey: \"%s\",error details: %s", referencedInstanceID, err)
+		return nil, err
+	}
+
+	// Ownership validation
+	if rif.tenantIdentifier != "" {
+		callerTenantID := query.RetrieveFromCriteria(rif.tenantIdentifier, query.CriteriaForContext(req.Context())...)
+		err = storage.ValidateOwnership(rif.repository, rif.tenantIdentifier, req, callerTenantID)
+		if err != nil {
+			if err == util.ErrNotFoundInStorage {
+				return nil, util.HandleInstanceSharingError(util.ErrMissingOrInvalidReferenceParameter, instance_sharing.ReferencedInstanceIDKey)
+			}
+			log.C(ctx).Errorf("Unable to Validate Ownership: \"%s\",error details: %s", rif.tenantIdentifier, err)
+			return nil, err
+		}
+	}
+
+	_, err = storage.IsReferencedShared(ctx, rif.repository, referencedInstanceID.String())
+	if err != nil {
+		log.C(ctx).Errorf("Failed to check if the referenced InstanceID is shared: \"%s\",error details: %s", referencedInstanceID, err)
+		// handled by the IsReferencedShared function
+		return nil, err
+	}
```

Comment:
this flow loads the instance twice from the DB, once for the ownership check and then again in IsReferenceShared. In addition, ValidateOwnership parses the body again to get the referenced instance ID, you can just pass it in. The same is correct for the plugin. If you're already there, take a look at the plugin and filter functions - they seem quite similar, maybe you can unite them, or at least extract some more logic to be common instead of implementing it twice.